docs/dev.md

```diff
@@ -0,0 +1,143 @@
+changes, and for bisecting for regressions. Each commit must build successfully,
+and should pass all tests, this may mean marking some tests as expected failures.
+
+A git commit should have a descriptive subject line, between 50 and 72
+characters in length. If it is not possible to be under 72 be as close as possible.
+It is helpful to preface the subject with a short description of what is
+affected. For example:
+
+```
+meson: bump gtest requirement from 1.12 to 1.15
+```
+
+After the subject should come a more detailed description of the change. This
+should include the reason for the change. It may also include implementation
+details, design decisions, and issues that may not be resolved or may need to be
+resolved in the future. Lines in this section should be wrapped to 72
+characters, or if this is not possible (such as a string that is itself longer
+than 72 characters) then be as close as possible.
+
+For example:
+
+```
+meson: bumpt gtest requirement from 1.12 to 1.15
+
+gtest 1.13 fixes a bug on Windows that is affecting our CI, see
+google/googletest#1234.
+
+gtest 1.14 adds a new feature that allows for faster test execution that will be
+used in the next patch
+```
+
+Finally, if a commit fixes any github issues, they should add that,
+
+```
+Fixes: #7
+```
+
+### Merging and rebasing
+
+Rebasing should generally be preferred to merges. Rebases create a linear
+history that is easy to bisect over, and easy for a human to understand. In a
+project like cps-config that lacks large subsystems this is easy to maintain.
+
+Merges may be appropriate for large additions or refactors that have moved
+quickly and may not have a clean history. This should generally be unnecessary.
+Merges also do not go through CI, and as such it's possible to create a state
+where main before the merge builds successfully, and the topic branch before the
+merge builds successfully, but the merge commit itself is broken.
+
+Back merges (merging the main branch back into a topic branch) should never
+happen. This makes bisects particularly hard and the history very messy. A
+rebase should be done in this case
+
+## Editors
+
+Use [editorconfig](https://editorconfig.org). Many editors support editorconfig
+natively, others have plugins. Otherwise, configure your editor to use the same
+settings.
+
+## All files
+
+All files should have an [SPDX license
+header](https://spdx.dev/learn/handling-license-info/), along with the relevant
+copyright holder information, as long as they have a comment syntax. JSON files
+do not, and as such should not have such a header.
+
+For example, in a C or C++ file:
+```cpp
+// SPDX-License-Identifier: MIT
+// Copyright 2024 Copyright Holder
+```
+
+## Meson
+
+Use the following naming conventions:
+
+- external_dependencies (found by `dependency()` or `compiler.find_library()`)
+  should be prefixed with `dep_`, ex: `dep_fmt`
+- internal dependencies (`declare_dependency()`) should be prefixed with `idep_`
```

Comment:
Just to catch me up... why? In my mind it's cleaner for each unit of code, even subdirectories, to just treat dependencies like dependencies.

Reply:
I prefer to use the `idep_` identifier in part for historical reasons. But probably doesn't make sense anymore, so I'll change that.

If you're into history, read on:

In ye old days, a consumer of a subproject would make as series of calls like this, pull a named variable from a subproject into their project:
```meson
sub_gtest = subproject('gtest')
dep_gtest = sub_gtest.get_variable('dep_gtest')
```

In the modern approach, a producer does something like:
```meson
dep_gtest = declare_dependency(...)
meosn.override_dependency('gtest', dep_gtest)
```
and then the consumer just does:
```meson
dep_gtest = dependency('gtest')
```
and Meson will do the right thing.